src/shogun/base/subset_iterators.h

```diff
@@ -0,0 +1,185 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_SUBSET_ITERATORS_H
+#define SHOGUN_SUBSET_ITERATORS_H
+
+#include <shogun/lib/SGVector.h>
+#include <shogun/mathematics/Math.h>
+
+namespace shogun
+{
+	template <typename T>
+	class CDenseFeatures;
+	class CDenseLabels;
+
+	class CFeatures;
+	class CLabels;
+
+	namespace subset_iterator_detail
+	{
+		template <typename T>
+		struct remove_cvptr
+		{
+			typedef std::remove_cv_t<std::remove_pointer_t<T>> type;
+		};
+
+		template <typename T>
+		using remove_cvptr_t = typename remove_cvptr<T>::type;
+
+		template <typename Container, typename T>
+		struct return_type_subset_iterator
+		{
+		};
+
+		template <typename T1, typename T2>
+		struct return_type_subset_iterator<CDenseFeatures<T1>, T2>
+		{
+			using type = SGVector<T1>;
+		};
+
+		template <typename T>
+		struct return_type_subset_iterator<CDenseLabels, T>
+		{
+			using type = float64_t;
+		};
+	} // namespace subset_iterator_detail
+
+	template <class IterableSubsetContainer, typename ST>
+	class SubsetIteratorBase
```

Comment:
need to come up with a better name now that the subset doesn't change the behaviour

Reply:
We can also do that later ;)